components/docker/container/component.go

```diff
@@ -3,83 +3,137 @@ package container
+	docker_client "github.com/docker/docker/client"
+
 	"github.com/leptonai/gpud/components"
 	docker_container_id "github.com/leptonai/gpud/components/docker/container/id"
+	pkg_file "github.com/leptonai/gpud/pkg/file"
+	components_metrics "github.com/leptonai/gpud/pkg/gpud-metrics"
 	"github.com/leptonai/gpud/pkg/log"
-	"github.com/leptonai/gpud/pkg/query"
+	"github.com/leptonai/gpud/pkg/process"
 )
 
-func New(ctx context.Context, cfg Config) components.Component {
-	cfg.Query.SetDefaultsIfNotSet()
-	setDefaultPoller(cfg)
-
-	cctx, ccancel := context.WithCancel(ctx)
-	getDefaultPoller().Start(cctx, cfg.Query, docker_container_id.Name)
-
-	return &component{
-		rootCtx: ctx,
-		cancel:  ccancel,
-		poller:  getDefaultPoller(),
-		cfg:     cfg,
+func New(ctx context.Context, ignoreConnectionErrors bool) components.Component {
+	cctx, cancel := context.WithCancel(ctx)
+	c := &component{
+		ctx:                    cctx,
+		cancel:                 cancel,
+		ignoreConnectionErrors: ignoreConnectionErrors,
 	}
+	return c
 }
 
 var _ components.Component = (*component)(nil)
 
 type component struct {
-	rootCtx context.Context
-	cancel  context.CancelFunc
-	poller  query.Poller
-	cfg     Config
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// In case the docker daemon is not running, we ignore such errors as
+	// 'Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?'.
+	ignoreConnectionErrors bool
+
+	lastMu   sync.RWMutex
+	lastData Data
 }
 
 func (c *component) Name() string { return docker_container_id.Name }
 
-func (c *component) Start() error { return nil }
+func (c *component) Start() error {
+	go func() {
+		ticker := time.NewTicker(1)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+				ticker.Reset(time.Minute)
+			}
+
+			c.CheckOnce()
+		}
+	}()
+	return nil
+}
 
 func (c *component) States(ctx context.Context) ([]components.State, error) {
-	last, err := c.poller.Last()
-	if err == query.ErrNoData { // no data
-		log.Logger.Debugw("nothing found in last state (no data collected yet)", "component", docker_container_id.Name)
-		return []components.State{
-			{
-				Name:    docker_container_id.Name,
-				Healthy: true,
-				Reason:  query.ErrNoData.Error(),
-			},
-		}, nil
+	c.lastMu.RLock()
+	lastData := c.lastData
+	c.lastMu.RUnlock()
+	return lastData.getStates(c.ignoreConnectionErrors)
+}
+
+type Data struct {
+	DockerPidFound bool              `json:"docker_pid_found,omitempty"`
+	Containers     []DockerContainer `json:"containers,omitempty"`
+
+	// timestamp of the last check
+	ts time.Time `json:"-"`
+	// error from the last check
+	err error `json:"-"`
+	// set to true if the error is the connection error to the docker daemon
+	connErr bool `json:"-"`
+}
+
+func (d Data) marshalJSON() ([]byte, error) {
+	return json.Marshal(d)
+}
+
+func (d Data) DescribeReason() string {
```

Comment:
All of these should be updated referring to the comments in containerd/pod/component.go